src/dune_engine/action.ml

```diff
@@ -154,16 +154,23 @@ let fold_one_step t ~init:acc ~f =
 
 include Action_mapper.Make (Ast) (Ast)
 
+(* TODO change [chdir] to use [Path.Build.t] in the directory. This will be
+   easier once [Action.t] is split off from the dune lang's action *)
 let chdirs =
   let rec loop acc t =
     let acc =
       match t with
-      | Chdir (dir, _) -> Path.Set.add acc dir
+      | Chdir (dir, _) -> (
+        match Path.as_in_build_dir dir with
+        | None ->
+          Code_error.raise "chdir ouside the build directory"
```

Comment:
Actually, could this be a user error too?

Reply:
We have validation on the dune lang side so this case isn't possible.

Reply:
:+1: